osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/test/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/firmware/firmwarefile/FirmwareFileTest.java

```diff
@@ -0,0 +1,132 @@
+  }
+
+  @Test
+  public void testHeader() throws UnsupportedEncodingException {
+    final FirmwareFile firmwareFile = new FirmwareFile(byteArray);
+    final FirmwareFileHeader header = firmwareFile.getHeader();
+    log.debug(header.toString());
+    assertThat(header.getFirmwareImageMagicNumberHex())
+        .isEqualTo(FirmwareFile.FIRMWARE_IMAGE_MAGIC_NUMBER);
+    assertThat(header.getHeaderVersionInt()).isEqualTo(0);
+    assertThat(header.getHeaderLengthInt()).isEqualTo(35);
+    assertThat(header.getFirmwareImageVersionHex()).isEqualTo("11004000");
+    assertThat(header.getFirmwareImageLengthInt()).isEqualTo(161456);
+    assertThat(header.getSecurityLengthInt()).isEqualTo(16);
+    assertThat(header.getSecurityTypeEnum()).isEqualTo(SecurityType.GMAC);
+    assertThat(header.getAddressLengthInt()).isEqualTo(8);
+    assertThat(header.getAddressTypeEnum()).isEqualTo(AddressType.MBUS_ADDRESS);
+    assertThat(header.getMbusDeviceSerialNumber()).isEqualTo("ffffffff");
+    assertThat(header.getMbusManufacturerId().getIdentification()).isEqualTo("GWI");
+    assertThat(header.getMbusVersionInt()).isEqualTo(80);
+    assertThat(header.getMbusDeviceType()).isEqualTo(DeviceType.GAS);
+    assertThat(header.getActivationTypeEnum())
+        .isEqualTo(ActivationType.MASTER_INITIATED_ACTIVATION);
+    assertThat(header.getActivationTimeHex()).isEqualTo("000000000000");
+  }
+
+  @Test
+  public void testFittingMbusDeviceSerialNumber() throws IOException, ProtocolAdapterException {
+    final FirmwareFile firmwareFile = this.createPartialWildcardFirmwareFile();
+
+    final String idHex = "0000FFFF";
+    final int noncompliantMbusDeviceSerialNumberInt = Integer.parseInt(idHex, 16);
+
+    assertDoesNotThrow(
+        () -> firmwareFile.setMbusDeviceSerialNumber(noncompliantMbusDeviceSerialNumberInt));
+  }
+
+  @Test
+  public void testMisfitMbusDeviceSerialNumber() throws IOException, ProtocolAdapterException {
+    final FirmwareFile firmwareFile = this.createPartialWildcardFirmwareFile();
+
+    final String idHex = "00010000";
+    final int noncompliantMbusDeviceSerialNumberInt = Integer.parseInt(idHex, 16);
+    final Exception exception =
+        assertThrows(
+            ProtocolAdapterException.class,
+            () -> {
+              firmwareFile.setMbusDeviceSerialNumber(noncompliantMbusDeviceSerialNumberInt);
+            });
+    assertThat(exception)
+        .hasMessage(
+            "MbusDevice Serial Number (%s) does not fit the range of serial numbers supported by this Firmware File (%s)",
+            idHex,
+            new StringBuffer(firmwareFile.getHeader().getMbusDeviceSerialNumber())
+                .reverse()
+                .toString());
+  }
+
+  private FirmwareFile createPartialWildcardFirmwareFile() {
+    // Changed fully wildcarded (FFFFFFFF) MbusDeviceSerialNumber in header with partially
+    // wildcarded one (FFFF0000, LSB first)
```

Comment:
For a bit more confidence in the tests, I think some extra tests or a more random-like pattern would be good.
Only using a pattern with repeated characters misses a bug I believe.
Since all the issues with the wildcard handling maybe even low values (lots of zeros) high values (lots of nines) and more random values.

This is an example (needs clean up before use, this has multiple checks in a single test for instance, which is not a good practice to keep in the code):
```java
  @Test
  void rejectsIdentificationNumberThatDoesNotMatchLessRegularPattern() {

    // identifications 98470000 through 98479999
    // pattern FFFF4798
    // invalid "98402417"
    // valid "98478634"

    final byte[] clonedByteArray = byteArray.clone();
    clonedByteArray[22] = (byte) 255;
    clonedByteArray[23] = (byte) 255;
    clonedByteArray[24] = (byte) 0x47;
    clonedByteArray[25] = (byte) 0x98;

    final FirmwareFile firmwareFile1 = new FirmwareFile(clonedByteArray.clone());
    System.out.println(firmwareFile1.getHeader().getMbusDeviceSerialNumber());

    String idHex = "98478634"; // within range 98470000 through 98479999
    final int compliantMbusIdentificationNumberInt = (int) Long.parseLong(idHex, 16);

    assertDoesNotThrow(
        () -> firmwareFile1.setMbusDeviceSerialNumber(compliantMbusIdentificationNumberInt));

    final FirmwareFile firmwareFile2 = new FirmwareFile(clonedByteArray.clone());
    System.out.println(firmwareFile2.getHeader().getMbusDeviceSerialNumber());
    idHex = "80215922"; // outside range 98470000 through 98479999
    final int noncompliantMbusDeviceSerialNumberInt = (int) Long.parseLong(idHex, 16);
    final Exception exception =
        assertThrows(
            ProtocolAdapterException.class,
            () -> {
              firmwareFile2.setMbusDeviceSerialNumber(noncompliantMbusDeviceSerialNumberInt);
            });
    assertThat(exception)
        .hasMessage(
            "MbusDevice Serial Number (%s) does not fit the range of serial numbers supported by this Firmware File (%s)",
            idHex, "9847FFFF");
  }
```

I am not sure what the best way is for the pattern to appear in the message. The spec has capital F's so that may be nicer than little f's as was the case.

Here is an example implementation of `checkWildcard` in `FirmwareFile` (not cleaned up, just coded till it worked) that would make the above test pass:
```java
  private void checkWildcard(final String mbusDeviceSerialNumberHex)
      throws ProtocolAdapterException {
    final String lsbFirstPattern = this.getHeader().getMbusDeviceSerialNumber();
    // lsbFirstPattern should have an even number of characters (a pair per byte)
    final int numberOfPairs = lsbFirstPattern.length() / 2;
    final StringBuilder patternBuilder = new StringBuilder();
    for (int i = 0; i < numberOfPairs; i++) {
      final int positionAfterPairToCopy = 2 * (numberOfPairs - i);
      patternBuilder
          .append(lsbFirstPattern.charAt(positionAfterPairToCopy - 2))
          .append(lsbFirstPattern.charAt(positionAfterPairToCopy - 1));
    }
    final String msbFirstPattern = patternBuilder.toString();
    final String patternRegex = msbFirstPattern.replaceAll("[fF]", "[0-9]");
    if (!Pattern.matches(patternRegex, mbusDeviceSerialNumberHex)) {
      throw new ProtocolAdapterException(
          String.format(
              "MbusDevice Serial Number (%s) does not fit the range of serial numbers supported by this Firmware File (%s)",
              mbusDeviceSerialNumberHex, msbFirstPattern.toUpperCase(Locale.UK)));
    }
  }
```

Note that this example only allows digits in the string representation of the identification number, and not `[a-fA-F]`.

Reply:
Added two tests with Less Regular Patterns. Added a separate method for reversing the hexadecimal string